src/Messenger/KafkaTransport.php

```diff
@@ -102,6 +103,22 @@ public function get(): iterable
         return [];
     }
 
+    private function getEnvelope(Message $message): iterable
+    {
+        try {
+            $envelope = $this->serializer->decode([
+                'body' => $message->payload,
+                'headers' => $message->headers,
+            ]);
+        } catch (MessageDecodingFailedException $exception) {
+            $this->rejectKafkaMessage($message);
```

Comment:
@KonstantinCodes this is also done in other transports like RabbitMQ, all enqueue integrations.

The question is now: if serialization fails what should it do? Stop the consumer everytime and try serialization the message everytime (untill the code is fixed to accept this message)?

Or do what it does now: try to deserialize and if it fails throw this exception in the log and ack the message (so no hammering will occur).

Reply:
The serializer throws an exception, either because of a programming error or because of an unexpected value. Maybe a new schema, maybe it's an avro message and the schema registry is temporarily down.

In that case, no messages will get processed but they will all be acked.

Once the schema registry is back online, messages will be processed again. We may just have lost a few hours of data. And the programmers have to manually reset the group.id to the last offset before the schema registry went offline, if they want to process the "lost" messages.

It is a very messy problem to fix.

On the other hand - if message processing waits until the schema registry has come online again, no manual intervention by the programmers is necessary. The system "healed" itself.

Reply:
Thanks for your reply! 

Im still doubting what to do about the failing serialization, it could have multiple issues:
- Registered schema does not completely match with code so errors could occur durring deserialization (logic thing that this should be fixed)
- There is no registered schema so 'anything' can be send on producer side and consume can fail
- Avro message and schema registry is down

By default the Symfony messenger setup is done with supervisord, so when a consumer stops it just restarts so when a message failed it can just consume the next if you setup the retry strategy correctly.

On all projects at the moment we do now;
- When serialization fails -> stop consumer and retry (untill now this doesn't occured yet)
- When consuming fails (during handling) -> move to failed db transport (able to retry)

But I see a problem arrises when serialization will fail for a single message and hold all the other thousands of messages in Kafka. Of course there is also a problem if all messages fails during serialization (because of down time schema registry or an incorrect code update on producer/consuming side).

Reply:
https://github.com/symfony/symfony-docs/pull/13597/files
This PR looks rather interesting. There is a bash script that sleeps for 30 seconds to prevent hammering.
Not sure if this could help but maybe worth a try?

Reply:
Yeah sorry, I didn't see your reply yet so I just edited my message https://github.com/KonstantinCodes/messenger-kafka/pull/26#discussion_r521237276 
Basically I'd try to catch the exception in the serializer and move the message the the failed bus.

Reply:
@KonstantinCodes indeed true. Good thing to check out consuming and supervisord :+1: 

Idea to change this PR to not ack when serialization fails (logic should then be in serialization class if it should do something and if it fails it will be retried?) and ack the message if it during handling fails and after that the Kafka can be recreated through retry strategy (eg move to a queue where you can delay a message for retrying like SQS/db/Redis/RabbitMQ) or to the bottom of the Kafka topic (if this is enabled via config)

+ set default ack when fail to false